Automattic/jetpack
# Tiled Gallery: Increasing accessibility of gallery images with respect to opening carousels

Fixes #28291

## Proposed changes:

* This PR adds a tabIndex of 0 and the role of button to images added via the Tiled Gallery block (and to Tiled Galleries added using the Classic Editor) to aid in ensuring the images can be focusable.
  * Note that choosing 'Attachment Page' or 'Media File' instead of 'None' as the 'Link To' option is actually a workaround in order to focus on the image using a screenreader / keyboard, but only if using full-screen carousels instead, as we are doing here.
* For Tiled Gallery blocks, an aria-label has also been added to the published gallery images - previously this only existed in the editor. For both Classic and block Tiled Gallery's the aria-label exists in the carousel image already. Adding the aria-label to the image in the gallery itself in the Classic Editor proved difficult, so due to this being likely less used than the block version I opted not to dig in too much more here.
*  It also updated the carousel JavaScript to listen out for more than just a click event in order to open the carousel - now it listens for 'keydown' (space or enter) (when the 'Link To' option is set to 'None', otherwise enter will work). Touch events are already picked up.
* This is in order to ensure that the carousel can be triggered when using the keyboard or a screen reader for navigation.
* I've not added `aria-controls` or `aria-expanded` - we'd be better off with something like a toggle to open up the carousel full-screen mode in that case and that would involve a redesign which wouldn't be a priority here.

### Other information:


## Jetpack product discussion

https://github.com/Automattic/jetpack/issues/28291

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

To replicate the issue:

* On trunk, enable the 'Display images in a full-screen carousel gallery' feature - `/wp-admin/admin.php?page=jetpack#/writing`
* On a post or page add a Tiled Gallery block, and publish the post/page. Ensure that the Tiled Gallery block is set to link to 'None' from the settings sidebar (note that choosing Attachment Page or Media File is actually a workaround in order to focus on the image using a screenreader / keyboard)
* On the front-end, click on a gallery image, and the carousel should open.
* Close the carousel, and use a keyboard to navigate to the images (tabbing for example). You'll notice you tab right past them (the images are missed)
* Similarly if you have access to a screen reader, attempt to navigate to the images, ad it won't be possible.
* To replicate the issue using the Classic Editor plugin, follow the steps here: https://jetpack.com/support/classic-tiled-galleries/. Make sure to select 'None' under 'Gallery Settings'->'Link To'

To test the fix:
* Follow the same steps as above. When using the keyboard or a screen-reader to navigate to the gallery images you should now be able to open the carousel when focussing on an image with either: 'enter', 'space', clicking or touching/tapping if that is an option for you.
* The same should apply to galleries created within the Classic Editor.
* Check the source for the aria-labels for images before opening a carousel (Tiled Gallery block).
